### DISCUSSION TITLE
Generalizing and standardizing the deletion-from-collection algorithm

### DISCUSSION THREAD

**Discussion by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#839] [Show and tell]
Hi guys,

I wanted to write some more tests of @BenjamenMeyer 's deletion-from-collection code. And also test a couple of other approaches, to see which one would be the best. You can see the results here:

https://github.com/stephengtuggy/raw_pointers_safe_deletion_tests/pull/1

Feedback is welcome. Thanks!

**Comment by BenjamenMeyer** [comment]
I'm all for improving. Interesting that you were able to find a bug there too, but glad to see it fixed.

If we can support C++14 without issue that'd probably be good. ISO guidelines are now being written for C++20 or newer (annoying they don't keep a historic record but such is life).

**Reply by stephengtuggy** [reply]
Yep. Sounds good.

@royfalk what are your thoughts?

**Comment by stephengtuggy** [comment]
https://www.fluentcpp.com/2018/09/18/how-to-remove-pointers-from-a-vector-in-cpp/

**Reply by BenjamenMeyer** [reply]
@stephengtuggy can we pass a function reference instead of a lambda? That would be better for testability since you could test the function directly outside of the deletion call.

Generally I'd avoid lambdas outside of the most simple cases. In this case:
```c++
vec.erase(
    std::remove_if(
        vec.begin(),
        vec.end(),
        [](int i){
            return (i % 2) == 0;
        }
    ),
    vec.end()
);
```
would become:
```c++
bool remove_test(int i) {
    return (i %2) == 0;
}
...
vec.erase(
    std::remove_if(
        vec.begin(),
        vec.end(),
        remove_test
    ),
    vec.end()
);
```

Thoughts?

**Reply by stephengtuggy** [reply]
I suppose that makes sense. There probably isn't any particular reason why it has to be a lambda, unless it's something to do with closures / capturing values or something like that.

**Reply by stephengtuggy** [reply]
Reviewing that fluentcpp.com article the other day, I realized that we really don't need to do anything special for collections of anything other than pointers. std::remove_if will work fine. We probably don't need to write a whole new function for that purpose.

Where writing a new delete function would come in handy is with collections of pointers -- especially owning raw pointers. I think it would be a good idea to standardize our approach to that.

**Reply by royfalk** [reply]
> Where writing a new delete function would come in handy is with collections of pointers -- especially owning raw pointers. I think it would be a good idea to standardize our approach to that.

That's easy, don't. Don't hold raw pointers in a collection without storing the pointed element somewhere. In other words, don't use new. That way, if you delete a pointer, nothing bad happens.

**Reply by stephengtuggy** [reply]
It's probably a good idea to switch everything over to `std::shared_ptr` and `std::weak_ptr`, etc. at some point. However, that will be a massive set of changes. It also raises some questions that remain unresolved, like what will trigger deletion of an object if a pointer to it is always held by at least one collection.

This proposal is basically intended as an interim measure.

**Reply by stephengtuggy** [reply]
FYI: I am working on the smart-pointer refactors, among other things, in parallel on branch https://github.com/stephengtuggy/Vega-Strike-Engine-Source/tree/pr-754_redo . This is an ongoing effort.

**Reply by BenjamenMeyer** [reply]
I'm in favor of using `std::shared_ptr` - used properly it handles the memory quite nicely so you can just allocate/deallocate an forget it since it'll do the actual cleanup when the reference counter hits zero.

**Reply by stephengtuggy** [reply]
> I'm in favor of using `std::shared_ptr` - used properly it handles the memory quite nicely so you can just allocate/deallocate an forget it since it'll do the actual cleanup when the reference counter hits zero.

Yes and no. If we use `std::shared_ptr` in the Unit collections, the reference counter on each will never hit zero; there will always be something that still has a reference to the Unit in question. I think we need a more sophisticated approach, probably one where we use `std::weak_ptr`s a lot, and have Units manage their own lifecycle, initiating their own destruction when `Kill()` is called or whatever.

**Reply by stephengtuggy** [reply]
Note that you don't need to use **both** vec.erase and std::remove_if. Just one of them. If you use both, you'll have problems, likely removing more elements from the collection than you intended to.

**Reply by royfalk** [reply]
We should probably store the units in a vector directly.

I understand this is a bit of refactoring but **this is the way**.

No *new*, no pointer, nothing.

When we're done, we erase a unit from the vector and it's gone.

**Reply by stephengtuggy** [reply]
@royfalk Okay. How would we refer to the units from other parts of the code then? By reference instead of by address?

**Reply by stephengtuggy** [reply]
And what if we wanted to add to the unit collection an instance of a Unit subclass, like Planet? How would that work?

**Reply by stephengtuggy** [reply]
We also need to figure out what to do about Meshes, Textures, Animations, Orders, and other classes, including their subclasses as applicable.

**Reply by royfalk** [reply]
We would refer to them as pointers in other parts of the engine. It's just that when deleting them, we would remove them from the vector, the destructor would be called and nothing would leak.

**Reply by stephengtuggy** [reply]
I see.

**Reply by BenjamenMeyer** [reply]
Avoiding new/delete is certainly a performance booster and is how Embedded applications manage memory - everything gets allocated on the stack or at program start so you don't dynamically allocate anything to keep things humming along without having to worry about memory allocation failures.

That said... I'm not sure. We could certainly have more stack-like allocations and utilize the C++ Move operator to assign things around - so a ship would have a list of components that would be managed as @royfalk suggests, and we could have different types of things along those lines. That could certainly be an interesting method of managing it all.

I'm open to the idea - I'd like to see how it'd work overall.

**Reply by stephengtuggy** [reply]
> We would refer to them as pointers in other parts of the engine. It's just that when deleting them, we would remove them from the vector, the destructor would be called and nothing would leak.

A couple of things. First, if we're going to be deleting items from the middle of the collection, then an `std::vector` might not be the best choice. Something like an `std::deque` might be more efficient.

Second: If we store objects directly in the collection, then they all have to have the same memory footprint. Any child classes inheriting from the base class cannot add any fields to the base, except for static fields. This is the way Meshes are stored today. And it's a bit tricky to work with.

**Reply by BenjamenMeyer** [reply]
yes, `std::deque` is far more efficient if you're randomly adding/removing things by many orders of magnitude over `std::vector` as `std::vector` allocates in blocks; if you need to expand the memory footprint then it reallocates and moves everything it contains; where as `std::deque` is more like a linked list/hashtable/b-tree style mechanism so things are really efficient.

**Comment by stephengtuggy** [comment]
I continue to see places in the code where removing items from a collection is being done wrong. Can we talk about this subject some more?
